File: Java/src/main/java/rental_system/Rentals.java
import java.io.*;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;

public class Rentals {
    private static final DateTimeFormatter DATE_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd");

    public static void initRentalsFile() {
        File file = new File("data/rentals.txt");
        if (!file.exists()) {
            try {
                Files.createDirectories(Paths.get("data"));
                file.createNewFile();
            } catch (IOException e) {
                System.out.println("Erreur lors de l'initialisation de rentals.txt : " + e.getMessage());
            }
        }
    }

    public static String rentItem(String item, String startDateStr, String endDateStr) {
        if (!Items.itemExists(item)) {
            return "Erreur : L'objet " + item + " n'existe pas.";
        }

        LocalDate startDate, endDate;
        try {
            startDate = LocalDate.parse(startDateStr, DATE_FORMATTER);
            endDate = LocalDate.parse(endDateStr, DATE_FORMATTER);
        } catch (DateTimeParseException e) {
            return "Erreur : Format de date invalide (utilisez YYYY-MM-DD).";
        }

        if (endDate.isBefore(startDate.plusDays(1))) {
            return "Erreur : La location doit durer au moins un jour.";
        }

        try {
            BufferedReader reader = new BufferedReader(new FileReader("data/rentals.txt"));
            String line;
            LocalDate now = LocalDate.now();
            while ((line = reader.readLine()) != null) {
                String[] parts = line.split(",");
                if (parts.length == 3 && parts[0].equals(item)) {
                    try {
                        LocalDate existingEnd = LocalDate.parse(parts[2], DATE_FORMATTER);
                        if (existingEnd.isAfter(now) || existingEnd.isEqual(now)) {
                            LocalDate existingStart = LocalDate.parse(parts[1], DATE_FORMATTER);
                            if (startDate.isBefore(existingEnd.plusDays(1)) && endDate.isAfter(existingStart.minusDays(1))) {
                                reader.close();
                                return "Erreur : " + item + " est déjà loué jusqu'à " + existingEnd + ".";
                            }
                        }
                    } catch (DateTimeParseException ignored) {
                    }
                }
            }
            reader.close();

            try (PrintWriter writer = new PrintWriter(new FileWriter("data/rentals.txt", true))) {
                writer.println(item + "," + startDateStr + "," + endDateStr);
            }
            return "Succès : " + item + " loué du " + startDateStr + " au " + endDateStr + ".";
        } catch (IOException e) {
            return "Erreur lors de l'ajout de la location : " + e.getMessage();
        }
    }

    public static void listRentals(String item) {
        if (!Items.itemExists(item)) {
            System.out.println("Erreur : L'objet " + item + " n'existe pas.");
            return;
        }

        boolean found = false;
        try (BufferedReader reader = new BufferedReader(new FileReader("data/rentals.txt"))) {
            String line;
            LocalDate now = LocalDate.now();
            while ((line = reader.readLine()) != null) {
                String[] parts = line.split(",");
                if (parts.length == 3 && parts[0].equals(item)) {
                    try {
                        LocalDate endDate = LocalDate.parse(parts[2], DATE_FORMATTER);
                        if (endDate.isAfter(now) || endDate.isEqual(now)) {
                            System.out.println("Loué du " + parts[1] + " au " + parts[2]);
                            found = true;
                        }
                    } catch (DateTimeParseException ignored) {
                    }
                }
            }
        } catch (IOException e) {
            System.out.println("Erreur lors de la lecture des locations : " + e.getMessage());
        }
        if (!found) {
            System.out.println("Aucune location active pour " + item + ".");
        }
    }
}

File: Go/rentals/rentals.go
package rentals

import (
	"fmt"
	"os"
	"strings"
	"time"
	"rental_system/items"
)

// Rental représente une location
type Rental struct {
	Item      string
	StartDate time.Time
	EndDate   time.Time
}

// InitRentalsFile crée rentals.txt s'il n'existe pas
func InitRentalsFile() error {
	if _, err := os.Stat("data/rentals.txt"); os.IsNotExist(err) {
		_, err := os.Create("data/rentals.txt")
		if err != nil {
			return fmt.Errorf("création de rentals.txt : %v", err)
		}
	}
	return nil
}

// RentItem ajoute une location après validation
func RentItem(item, startDateStr, endDateStr string) error {
	// Vérifier si l'objet existe
	exists, err := items.ItemExists(item)
	if err != nil {
		return err
	}
	if !exists {
		return fmt.Errorf("l'objet %s n'existe pas", item)
	}

	// Parser les dates
	startDate, err := time.Parse("2006-01-02", startDateStr)
	if err != nil {
		return fmt.Errorf("date de début invalide : %v", err)
	}
	endDate, err := time.Parse("2006-01-02", endDateStr)
	if err != nil {
		return fmt.Errorf("date de fin invalide : %v", err)
	}

	// Vérifier la durée minimale (1 jour)
	if endDate.Before(startDate.Add(24 * time.Hour)) {
		return fmt.Errorf("la location doit durer au moins un jour")
	}

	// Vérifier les conflits
	rentals, err := loadRentals()
	if err != nil {
		return err
	}
	now := time.Now()
	for _, r := range rentals {
		if r.Item == item && r.EndDate.After(now) {
			if startDate.Before(r.EndDate) && endDate.After(r.StartDate) {
				return fmt.Errorf("%s est déjà loué jusqu'à %s", item, r.EndDate.Format("2006-01-02"))
			}
		}
	}

	// Ajouter la location
	file, err := os.OpenFile("data/rentals.txt", os.O_APPEND|os.O_WRONLY, 0644)
	if err != nil {
		return fmt.Errorf("ouverture de rentals.txt : %v", err)
	}
	defer file.Close()
	_, err = file.WriteString(fmt.Sprintf("%s,%s,%s\n", item, startDateStr, endDateStr))
	if err != nil {
		return fmt.Errorf("écriture dans rentals.txt : %v", err)
	}
	return nil
}

// ListRentals affiche les locations actives d'un objet
func ListRentals(item string) error {
	exists, err := items.ItemExists(item)
	if err != nil {
		return err
	}
	if !exists {
		return fmt.Errorf("l'objet %s n'existe pas", item)
	}

	rentals, err := loadRentals()
	if err != nil {
		return err
	}

	now := time.Now()
	found := false
	for _, r := range rentals {
		if r.Item == item && r.EndDate.After(now) {
			fmt.Printf("Loué du %s au %s\n", r.StartDate.Format("2006-01-02"), r.EndDate.Format("2006-01-02"))
			found = true
		}
	}
	if !found {
		fmt.Printf("Aucune location active pour %s.\n", item)
	}
	return nil
}

// loadRentals charge les locations depuis rentals.txt
func loadRentals() ([]Rental, error) {
	data, err := os.ReadFile("data/rentals.txt")
	if err != nil {
		return nil, fmt.Errorf("lecture de rentals.txt : %v", err)
	}
	var rentals []Rental
	lines := strings.Split(strings.TrimSpace(string(data)), "\n")
	for _, line := range lines {
		if line == "" {
			continue
		}
		parts := strings.Split(line, ",")
		if len(parts) != 3 {
			continue
		}
		startDate, err := time.Parse("2006-01-02", parts[1])
		if err != nil {
			continue
		}
		endDate, err := time.Parse("2006-01-02", parts[2])
		if err != nil {
			continue
		}
		rentals = append(rentals, Rental{
			Item:      parts[0],
			StartDate: startDate,
			EndDate:   endDate,
		})
	}
	return rentals, nil
}